objects/src/notes/inputs.rs

```diff
@@ -0,0 +1,50 @@
+use super::{Digest, Felt, Hasher, NoteError, ZERO};
+
+/// Holds the inputs which are placed onto the stack before a note's script is executed.
+#[derive(Debug, Eq, PartialEq)]
+pub struct NoteInputs {
+    inputs: Vec<Felt>,
+    hash: Digest,
+}
+
+impl NoteInputs {
+    /// Number of note inputs.
+    const NOTE_NUM_INPUTS: usize = 16;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns NoteInputs created from the provided inputs.
+    ///
+    /// The inputs are padded with ZERO such that they are always of length 16.
+    ///
+    /// # Errors
+    /// Returns an error if the number of provided inputs is greater than 16.
+    pub fn new(inputs: &[Felt]) -> Self {
+        if inputs.len() > Self::NOTE_NUM_INPUTS {
+            NoteError::too_many_inputs(inputs.len());
+        }
+
+        // pad inputs with ZERO to be constant size (16 elements)
+        let mut padded_inputs = Vec::with_capacity(Self::NOTE_NUM_INPUTS);
+        padded_inputs.extend_from_slice(inputs);
+        padded_inputs.resize(Self::NOTE_NUM_INPUTS, ZERO);
```

Comment:
I wonder if this is the right approach to padding - especially in how we arrange inputs on the stack. For example, if the inputs are provided as `[a, b]`, I'd expect them to be located on the stack as `[b, a, ...]` (first input deep in the stack). But if we pad them the way they are padded here, they'd be pushed onto the stack as `[0, 0, 0 ..., 0, b, a, ...]`.

Reply:
good catch. I've modified the logic such that we pad, compute hash in stack order and then reverse them before we instantiate the NoteInputs struct.